File: day03/challenge2.py
#!/usr/bin/env python3
import sys
import re

def main():
    #read input from stdin
    lines = "".join([line.strip() for line in sys.stdin])

    #a regular expression that captures every instance of mul() with two 1-3 digit numbers as arguments
    valid_exp = r"(mul\(\d{1,3}\s*,\s*\d{1,3}\)|don\'t\(\)|do\(\))"

    #a regular expression that extracts the two numeric arguments to mul()
    mul_exp = r"mul\(\s*(\d{1,3})\s*,\s*(\d{1,3})"

    #a regular expression to remove any instructions between a don't() and a do()
    disable_exp = r"(don\'t\(.*\).*?do\(.*?\))"
    
    print(lines)

    #match valid expressions
    expressions = [match.groups() for match in re.finditer(valid_exp, lines)]
    mul = True
    sum = 0
    for exp in expressions:
        print(exp[0])
        if exp[0] == "don't()":
            mul = False
        elif exp[0] == "do()":
            mul = True
        else:
            if not mul:
                continue
            match = re.match(mul_exp, exp[0])
            sum+= int(match.group(1)) * int(match.group(2))


    print(expressions)
    print(sum)

if __name__ == "__main__":
    main()

File: day03/challenge2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

func main() {
	//read input from stdin into an array of strings
	scanner := bufio.NewScanner(os.Stdin)
	var lines []string
	for scanner.Scan() {
		line := scanner.Text()
		lines = append(lines, line)
	}
	//fmt.Println("lines:", lines)
	//regular expression to match valid instructions
	//mul(\d{1,3}, \d{1,3})
	//do()
	//don't()
	var re = regexp.MustCompile(`mul\((\d{1,3}),(\d{1,3})\)|do\(\)|don\'t\(\)`)
	var mul = true
	var sum = int64(0)

	for _, line := range lines {
		fmt.Println("line:", line)
		//match the instruction
		var match = re.FindAllStringSubmatch(line, -1)
		//fmt.Println("match:", match)
		for _, m := range match {
			fmt.Println("match: ", m)
			if m[0] == "do()" {
				mul = true
			} else if m[0] == "don't()" {
				mul = false
			} else {
				if mul {
					a, err := strconv.ParseInt(m[1], 10, 64)
					if err != nil {
						fmt.Println("error:", err)
						return
					}
					b, err := strconv.ParseInt(m[2], 10, 64)
					if err != nil {
						fmt.Println("error:", err)
						return
					}
					sum += a * b
				}
			}
		}
	}
	fmt.Println("sum:", sum)
}
